### PR TITLE
Add parameters for qualified business income and update to taxcalc 3.0

### PR DESCRIPTION
**Pull request by Peter-Metz**
This PR adds ~five~ four parameters to Tax-Cruncher for calculation of the qualified business income deduction:
- ~`pbusinc`: Primary filer's business income~
- ~`sbusinc`: Spouse's business income~
- `businc`: Business income (see [comment](https://github.com/PSLmodels/Tax-Cruncher/pull/38#issuecomment-680952838))
- `sstb`: Boolean for whether business income is from a professional services business
- `w2paid`: W-2 wages paid by business
- `qualprop`: Filer's share of qualified property

Also, this PR brings Tax-Cruncher up to speed with Tax-Calculator 3.0. Notably, this simplifies parameter processing within `cs-config/` because of Tax-Calculator's adoption of Paramtools.

@hdoupe would you mind reviewing the changes related to CS?

@MattHJensen and @kpomerleau might be interested in the new QBID functionality

### CODE REVIEW

**Comment by Peter-Metz** [comment]
In the latest commits, I've consolidated `pbusinc` and `sbusinc` into `businc`. Previously, `pbusinc` translated to `e00900p` and `sbusinc` translated to `e00900s`. Now, `businc` translates to `e26270`.

The reason for this change is twofold:
1. Tax-Calculator uses `e26270` in [its test](https://github.com/PSLmodels/Tax-Calculator/blob/master/taxcalc/tests/test_calculator.py#L822) for QBID logic. I've adopted a very similar test for Tax-Cruncher that returns the same QBID amounts.
2. There is an input field in Tax-Cruncher, but no Tax-Calculator parameter, for passive S-corp income. However, by asking for passive and activate S-corp income separately in Tax-Cruncher, we can calculate total S-corp income (`e02000`) to pass to Tax-Calculator and avoid double counting.

File: cs-config/cs_config/functions.py [hunk 1]
```diff
@@ -20,11 +19,9 @@
 with open(os.path.join(TCDIR, "policy_current_law.json"), "r") as f:
     pcl = json.loads(f.read())
 
-RES = convert_defaults(pcl)
-
 
 class TCParams(paramtools.Parameters):
-    defaults = RES
+    defaults = pcl
```
Comment (hdoupe) [inline cs-config/cs_config/functions.py:24]: @Peter-Metz did you try using `taxclac.Policy` directly? That will have all of the more sophisticated validation relating to indexing.

Comment (Peter-Metz) [reply cs-config/cs_config/functions.py:24]: @hdoupe when I try that I get:

```
E           TypeError: Unable to read data of type: <class 'taxcalc.policy.Policy'>
E                      Data must be a File Path, URL, String, or Dict.

Comment (hdoupe) [reply cs-config/cs_config/functions.py:24]: Ah, I meant you can get rid of the `TCParams` class and just use the `Policy` class (with a couple modifications). 

The two tricky things are the extra `data_source` label that is used to filter inputs depending on whether the PUF or CPS is being used. This may not affect Tax-Cruncher though since it's just using the CPS. The other tricky thing will be converting `_checkbox` to `-index` when validating/processing the reforms for Tax-Calculator.

Happy to help work through both of these issues. 

(Now that I'm thinking about it, it may be easier to handle the PUF/CPS thing another way--but this may not matter for taxcrunch)

Comment (Peter-Metz) [reply cs-config/cs_config/functions.py:24]: Ah that makes sense. When I try replacing `TCParams` with `Policy`, I get a huge traceback that is hard to follow. Here is one part that may be relevant though:

```
self = <cs_kit.validate.ParamsPolicy object at 0x7ffaad53da50>, param = 'SS_Earnings_c'
lte_val = 2018

    def get_index_rate(self, param: str, lte_val: Any):
        """
        Return the value of the index_rates dictionary matching the
        label to extend value, `lte_val`.
        Projects may find it convenient to override this method with their own
        `index_rate` method.
        """
>       return self.index_rates[lte_val]
E       KeyError: 2018

Comment (hdoupe) [reply cs-config/cs_config/functions.py:24]: @Peter-Metz Can you post the full traceback?

Comment (Peter-Metz) [reply cs-config/cs_config/functions.py:24]: @hdoupe here ya go:

```
____________________________ TestFunctions1.test_get_inputs ____________________________

model_parameters = {'Policy': OrderedDict([('parameter_indexing_CPI_offset', {'validators': {'range': {'min': -0.005, 'max': 0.005}}, 'se...: 'Child Tax Credit (include children under 13)', 'notes': '', 'number_dims': 0, 'section_1': 'Dependents', ...}, ...}}

    def load_model_parameters(model_parameters):
        for sect, _defaults in model_parameters.items():
            try:
>               type(f"Params{sect}", (Parameters,), {"defaults": _defaults})()

../anaconda3/envs/taxcrunch-env/lib/python3.7/site-packages/cs_kit/validate.py:20: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cs_kit.validate.ParamsPolicy object at 0x7ffaad53da50>, initial_state = None
index_rates = None, ops = {}
schemafactory = <paramtools.schema_factory.SchemaFactory object at 0x7ffaad53dad0>
name = 'uses_extend_func'
v = <fields.Str(default=<marshmallow.missing>, attribute=None, validate=[<OneOf(choices=['0kids', '1kid', '2kids', '3+kids... 'validator_failed': 'Invalid value.', 'invalid': 'Not a valid string.', 'invalid_utf8': 'Not a valid utf-8 string.'})>
default_ops = [('array_first', False), ('label_to_extend', None), ('uses_extend_func', False)]
schema_ops = {'array_first': True, 'label_to_extend': 'year', 'uses_extend_func': True}
default = False, prev_array_first = True

    def __init__(
        self,
        initial_state: Optional[dict] = None,
        index_rates: Optional[dict] = None,
        **ops,
    ):
        schemafactory = SchemaFactory(self.defaults)
        (
            self._defaults_schema,
            self._validator_schema,
            self._schema,
            self._data,
        ) = schemafactory.schemas()
        self.label_validators = schemafactory.label_validators
        self._stateless_label_grid = OrderedDict()
        for name, v in self.label_validators.items():
            if hasattr(v, "grid"):
                self._stateless_label_grid[name] = v.grid()
            else:
                self._stateless_label_grid[name] = []
        self.label_grid = copy.deepcopy(self._stateless_label_grid)
        self._validator_schema.context["spec"] = self
        self._warnings = {}
        self._errors = {}
        self._state = self.parse_labels(**(initial_state or {}))
        self._search_trees = {}
        self.index_rates = index_rates or self.index_rates
    
        # set operators in order of importance:
        # __init__ arg: most important
        # class attribute: middle importance
        # schema action: least important
        # default value if three above are not specified.
        default_ops = [
            ("array_first", False),
            ("label_to_extend", None),
            ("uses_extend_func", False),
        ]
        schema_ops = self._schema.get("operators", {})
        for name, default in default_ops:
            if name in ops:
                setattr(self, name, ops.get(name))
            elif getattr(self, name, None) != default:
                setattr(self, name, getattr(self, name))
            elif name in schema_ops:
                setattr(self, name, schema_ops[name])
            else:
                setattr(self, name, default)
    
        if self.label_to_extend:
            prev_array_first = self.array_first
            self.array_first = False
            self.set_state()
>           self.extend()

../anaconda3/envs/taxcrunch-env/lib/python3.7/site-packages/paramtools/parameters.py:95: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cs_kit.validate.ParamsPolicy object at 0x7ffaad53da50>, label = 'year'
label_values = None, params = None, raise_errors = True, ignore_warnings = False

    def extend(
        self,
        label: Optional[str] = None,
        label_values: Optional[List[Any]] = None,
        params: Optional[List[str]] = None,
        raise_errors: bool = True,
        ignore_warnings: bool = False,
    ):
        """
        Extend parameters along `label`.
    
        **Parameters**
    
        - `label`: Label to extend values along. By default, `label_to_extend`
          is used.
        - `label_values`: values of `label` to extend. By default, this is a grid
          created from the valid values of `label_to_extend`.
        - `params`: Parameters to extend. By default, all parameters are extended.
        - `raise_errors`: Whether `adjust` should raise or store errors.
        - `ignore_warnings`: Whether `adjust` should raise or ignore warnings.
    
        **Raises**
    
          - `InconsistentLabelsException`: Value objects do not have consistent
            labels.
        """
        if label is None:
            label = self.label_to_extend
        else:
            label = label
    
        spec = self.specification(meta_data=True)
        if params is not None:
            spec = {
                param: self._data[param]
                for param, data in spec.items()
                if param in params
            }
        full_extend_grid = self._stateless_label_grid[label]
        if label_values is not None:
            labels = self.parse_labels(**{label: label_values})
            extend_grid = labels[label]
        else:
            extend_grid = self._stateless_label_grid[label]
    
        cmp_funcs = self.label_validators[label].cmp_funcs(choices=extend_grid)
        gt_cmp_func = make_cmp_func(cmp_funcs["gt"], all_or_any=all)
    
        adjustment = defaultdict(list)
        for param, data in spec.items():
            if not any(label in vo for vo in data["value"]):
                continue
            extended_vos = set()
            for vo in sorted(
                data["value"], key=lambda val: cmp_funcs["key"](val[label])
            ):
                hashable_vo = utils.hashable_value_object(vo)
                if hashable_vo in extended_vos:
                    continue
                else:
                    extended_vos.add(hashable_vo)
                gt = select(
                    self._data[param]["value"],
                    False,
                    gt_cmp_func,
                    {label: vo[label]},
                    tree=self._search_trees.get(param),
                )
                eq = select_eq(
                    gt,
                    False,
                    utils.filter_labels(vo, drop=["value", label, "_auto"]),
                )
                extended_vos.update(map(utils.hashable_value_object, eq))
                eq += [vo]
    
                defined_vals = {eq_vo[label] for eq_vo in eq}
    
                missing_vals = sorted(
                    set(extend_grid) - defined_vals, key=cmp_funcs["key"]
                )
    
                if not missing_vals:
                    continue
    
                extended = defaultdict(list)
                for vo in eq:
                    extended[vo[label]].append(vo)
    
                skl = utils.SortedKeyList(extended.keys(), cmp_funcs["key"])
    
                for val in missing_vals:
                    lte_val = skl.lte(val)
                    if lte_val is not None:
                        closest_val = lte_val
                    else:
                        closest_val = skl.gte(val)
    
                    if closest_val in extended:
                        value_objects = extended.pop(closest_val)
                    else:
                        value_objects = select_eq(
                            eq, False, {label: closest_val}
                        )
                    # In practice, value_objects has length one.
                    # Theoretically, there could be multiple if the inital value
                    # object had less labels than later value objects and thus
                    # matched multiple value objects.
                    for value_object in value_objects:
                        ext = dict(value_object, **{label: val})
                        ext = self.extend_func(
>                           param, ext, value_object, full_extend_grid, label
                        )

../anaconda3/envs/taxcrunch-env/lib/python3.7/site-packages/paramtools/parameters.py:781: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cs_kit.validate.ParamsPolicy object at 0x7ffaad53da50>, param = 'SS_Earnings_c'
extend_vo = {'value': 132900.0, 'year': 2013}
known_vo = {'value': 132900.0, 'year': 2019}
extend_grid = [2013, 2014, 2015, 2016, 2017, 2018, ...], label = 'year'

    def extend_func(
        self,
        param: str,
        extend_vo: ValueObject,
        known_vo: ValueObject,
        extend_grid: List,
        label: str,
    ):
        """
        Function for applying indexing rates to parameter values as they
        are extended. Projects may implement their own `extend_func` by
        overriding this one. Projects need to write their own `indexing_rate`
        method for returning the correct indexing rate for a given parameter
        and value of `label`.
    
        **Returns**
    
          - `extend_vo`: New `ValueObject`.
        """
        if not self.uses_extend_func or not self._data[param].get(
            "indexed", False
        ):
            return extend_vo
    
        known_val = known_vo[label]
        known_ix = extend_grid.index(known_val)
    
        toext_val = extend_vo[label]
        toext_ix = extend_grid.index(toext_val)
    
        if toext_ix > known_ix:
            # grow value according to the index rate supplied by the user defined
            # self.indexing_rate method.
            for ix in range(known_ix, toext_ix):
                v = extend_vo["value"] * (
                    1 + self.get_index_rate(param, extend_grid[ix])
                )
                extend_vo["value"] = np.round(v, 2) if v < 9e99 else 9e99
        else:
            # shrink value according to the index rate supplied by the user defined
            # self.indexing_rate method.
            for ix in reversed(range(toext_ix, known_ix)):
                v = (
                    extend_vo["value"]
>                   * (1 + self.get_index_rate(param, extend_grid[ix])) ** -1
                )

../anaconda3/envs/taxcrunch-env/lib/python3.7/site-packages/paramtools/parameters.py:843: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cs_kit.validate.ParamsPolicy object at 0x7ffaad53da50>, param = 'SS_Earnings_c'
lte_val = 2018

    def get_index_rate(self, param: str, lte_val: Any):
        """
        Return the value of the index_rates dictionary matching the
        label to extend value, `lte_val`.
        Projects may find it convenient to override this method with their own
        `index_rate` method.
        """
>       return self.index_rates[lte_val]
E       KeyError: 2018

../anaconda3/envs/taxcrunch-env/lib/python3.7/site-packages/paramtools/parameters.py:855: KeyError

The above exception was the direct cause of the following exception:

self = <cs_config.tests.test_functions.TestFunctions1 object at 0x7ffaad36c910>

    def test_get_inputs(self):
        self.test_all_data_specified()
        inputs = self.get_inputs({})
        check_get_inputs(inputs)
        init_metaparams = inputs["meta_parameters"]
        init_modparams = inputs["model_parameters"]
    
        try:
            json.dumps(init_metaparams)
        except TypeError as e:
            raise SerializationError(
                (
                    f"Meta parameters must be JSON serializable: \n\n\t{str(e)}\n"
                    f"\nHint: try setting `serializable=True` in `Parameters.specification`."
                )
            )
    
        try:
            json.dumps(init_modparams)
        except TypeError as e:
            raise SerializationError(
                (
                    f"Model parameters must be JSON serializable: \n\n\t{str(e)}\n"
                    f"\nHint: try setting `serializable=True` in `Parameters.specification`."
                )
            )
    
        class MetaParams(Parameters):
            array_first = True
            defaults = init_metaparams
    
        metaparams = MetaParams()
        assert metaparams
    
>       load_model_parameters(init_modparams)

../anaconda3/envs/taxcrunch-env/lib/python3.7/site-packages/cs_kit/validate.py:127: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

model_parameters = {'Policy': OrderedDict([('parameter_indexing_CPI_offset', {'validators': {'range': {'min': -0.005, 'max': 0.005}}, 'se...: 'Child Tax Credit (include children under 13)', 'notes': '', 'number_dims': 0, 'section_1': 'Dependents', ...}, ...}}

    def load_model_parameters(model_parameters):
        for sect, _defaults in model_parameters.items():
            try:
                type(f"Params{sect}", (Parameters,), {"defaults": _defaults})()
            except Exception as ser_exception:
                raise CSKitError(
                    f"An error was thrown while loading the model parameters for section: '{sect}'."
                    f"\n\n\t Hint: Make sure you are using the 'Parameters.dump' method if you are "
                    f"using ParamTools."
>               ) from ser_exception
E               cs_kit.exceptions.CSKitError: An error was thrown while loading the model parameters for section: 'Policy'.
E               
E               	 Hint: Make sure you are using the 'Parameters.dump' method if you are using ParamTools.

../anaconda3/envs/taxcrunch-env/lib/python3.7/site-packages/cs_kit/validate.py:26: CSKitError
=============================== short test summary info ================================
FAILED cs-config/cs_config/tests/test_functions.py::TestFunctions1::test_get_inputs

Comment (hdoupe) [reply cs-config/cs_config/functions.py:24]: @Peter-Metz Ha, no kidding about this stack trace...this is definitely something that can be improved. Also some error messages from the initial exception would be nice. I'll take a look at this tomorrow.

**Comment by Peter-Metz** [comment]
Thanks a lot for the assist @hdoupe! I'll merge this today if no one has objections
